### ISSUE TITLE
file.managed: UTF-8 in contents_pillar broken in 2018.3.3

### ISSUE THREAD

**Issue by tkaehn** [saltstack/salt#50219]
### Description of Issue/Question

UTF-8 in contents_pillar is broken in 2018.3.3. Before the last upgrade it worked.

I've also tried to set  `- encoding: utf-8` in the state file. But this didn't change the behaviour.

### Setup and steps to reproduce

```
root@minion:~# cat /srv/salt/motd.sls 
---
/etc/motd:
  file.managed:
    - contents_pillar: motd

root@minion:~# file /srv/pillar/file_tree/hosts/minion/motd
/srv/pillar/file_tree/hosts/minion/motd: UTF-8 Unicode text

root@minion:~# salt-call pillar.get motd
local:
    abcäöü

root@minion:~# salt-call state.apply motd
[ERROR   ] An exception occurred in this state: Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/state.py", line 1913, in call
    **cdata['kwargs'])
  File "/usr/lib/python2.7/dist-packages/salt/loader.py", line 1898, in wrapper
    return f(*args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/salt/states/file.py", line 2433, in managed
    [line.rstrip('\n').rstrip('\r') for line in validated_contents]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 3: ordinal not in range(128)

local:
----------
          ID: /etc/motd
    Function: file.managed
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/usr/lib/python2.7/dist-packages/salt/state.py", line 1913, in call
                  **cdata['kwargs'])
                File "/usr/lib/python2.7/dist-packages/salt/loader.py", line 1898, in wrapper
                  return f(*args, **kwargs)
                File "/usr/lib/python2.7/dist-packages/salt/states/file.py", line 2433, in managed
                  [line.rstrip('\n').rstrip('\r') for line in validated_contents]
              UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 3: ordinal not in range(128)
     Started: 08:24:29.933999
    Duration: 2.573 ms
     Changes:   

Summary for local
------------
Succeeded: 0
Failed:    1
------------
Total states run:     1
Total run time:   2.573 ms
```

### Versions Report

```
root@minion:~# salt-call --versions-report
Salt Version:
           Salt: 2018.3.3
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.5.3
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.9.4
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.13 (default, Sep 26 2018, 18:42:22)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.4.3
            ZMQ: 4.2.1
 
System Versions:
           dist: debian 9.5 
         locale: UTF-8
        machine: x86_64
        release: 4.9.0-7-amd64
         system: Linux
        version: debian 9.5 
```

**Comment by damon-atkins** [comment]
Also are utf8 however the output is determined by LANG environment vars

**Comment by Ch3LL** [comment]
i was able to replicate this and bisected it to this commit: https://github.com/saltstack/salt/pull/49316/commits/3fef112409ad206f4c6af5f21561dd71ba53b618

we will need to fix this up thanks

**Comment by rallytime** [comment]
@dwoz Can you take a look here?

**Comment by bornwitbugs** [comment]
@dwoz not sure if this is a bug or simply a packaging error. Ran into this same issue when using `file.blockreplace`. After playing around with the current code base (including the commit mentioned above by @Ch3LL ) I've found a discrepancy with the current source for salt/utils/files.py at least on centos 7 repo package.
When installing from salt repo `Check for ASCII first` is the first encoding check in the `get_encoding` function and starts at line #870 whereas in the master here on github `Check for ASCII first` is the last check. By manually fixing the order in my local version to match remote this issue seems to be resolved.

**Comment by dwoz** [comment]
os.linesep is a bytestring (not unicode) when running under python 2. It looks like that is causing an implicit encode call.

**Comment by dwoz** [comment]
@bornwitbugs are you looking at the 2018.3 branch and not the 2018.3.3 tag? That would explain the difference between what you see in the package vs the repository.

**Comment by dwoz** [comment]
I can confirm this has been fixed in the 2018.3 branch due to a combination of these PRs:

https://github.com/saltstack/salt/pull/49561

and 

https://github.com/saltstack/salt/pull/49782
